File: MaxAreaOfIsland.java
/**
 * LeetCode 695 Max Area of Island
 */
public int maxAreaOfIsland(int[][] grid) {
        if (grid == null) {
            return 0;
        }
        int r = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {
                if (grid[i][j] == 1) {
                    int num = search(grid, i, j);
                    r = Math.max(r, num);
                }
            }
        }
        return r;
    }

    public int search(int[][] grid,int i,int j) {
        if (i >= 0 && i < grid.length && j >= 0 && j < grid[i].length && grid[i][j] == 1) {
            //no repeat
            grid[i][j] = 0;
            int num = 1 + search(grid,i + 1,j) + search(grid , i - 1,j) + search(grid,i,j + 1) + search(grid,i, j - 1);
            return num;
        } else {
            return 0;
        }
    }

File: MaxAreaOfIsLand.go
var dire = []int{-1,0,1,0,-1}

func maxAreaOfIsland(grid [][]int) int {
	if (grid == nil || len(grid) == 0) {
		return 0
	}
	maxArea := 0
	for i := 0;i < len(grid);i++ {
		for j := 0;j < len(grid[i]);j++ {
			if (grid[i][j] == 1) {
				maxArea = max(maxArea,dfs(grid,i,j))
			}
		}
	}
	return maxArea
}

func dfs(grid [][]int,r int ,c int) int {
	grid[r][c] = 0
	x, y, area := 1,1,1
	for i := 0;i < 4;i++ {
		x = r + dire[i]
        y = c + dire[i + 1]
		if (x >= 0 && x < len(grid) && y >= 0 && y < len(grid[r]) && grid[x][y] == 1) {
			area += dfs(grid,x,y)
		}
	}
	return area
}

func max(a int,b int) int {
	if (a > b) {
		return a
	}
	return b
}
